<a href="https://colab.research.google.com/github/enikolaev/IT-and-Programming/blob/main/%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B0%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практическая работа 7. Обработка данных из файла. Использование коллекций

ФИО: **Николаев Евгений Иванович**

Группа: **ИНС-б-о-...**

## Задание

На рисунке показан фрагмент файла `data_v3.csv`, содержащего информацию о держателях финансовых карт.

<figure>
<center>
<img src='https://raw.githubusercontent.com/enikolaev/IT-and-Programming/refs/heads/main/data/image_for_pract_07.png'/>
<figcaption>Фрагмент csv-файла в github</figcaption></center>
</figure>

В data_v3.csv присутствуют следующие поля: `ID` – уникальный ключ записи; `LastName`, `firstName` – фамилия и имя держателя карты; `email` – электронная почта; `Plus` и `Minus` – суммы поступлений и списаний по карте соответственно в долларах; `credit_card` – номер карты; `Tax` – уплаченный налог.
Необходимо ответить на следующий вопросы:
1. Найдите держателя карты с максимальной прибылью. В ответе необходимо указать ФИО держателя и максимальную прибыль (прибыль = поступления – списания).
2. Найдите количество держателей карт, у которых сумма уплаченного налога больше среднего арифметического значения налога по всей выборке.


## Анализ задания, получение данных

Файл data_v3.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [1]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv


--2025-01-16 20:13:34--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 92780 (91K) [text/plain]
Saving to: ‘data_v3.csv’

data_v3.csv         100%[===================>]  90.61K  --.-KB/s    in 0.02s   

2025-01-16 20:13:34 (3.56 MB/s) - ‘data_v3.csv’ saved [92780/92780]



Можно скопировать и поместить файл с данными любым удобным способом

Реализуем чтение всего файла построчно. Используем метод `open()`, а затем будем обходить файл как коллекцию с использованием цикла `for line in File`. Очевидно, что csv-файл используется в данной работе как обычный текстовый файл и в цикле `for` производится последовательный обход всех строк этого файла, то есть текстовый файл - это коллекция его строк.
В файле много строк, поэтому будем выводить только первые `n_to_print` строк, но обрабатывать будем все (чтобы не возникло скрытых ошибок)

In [2]:
F=open('data_v3.csv')
n_to_print=10
for line in F:
  n_to_print-=1
  if n_to_print>0:
    print(line)

ID,LastName,FirstName,email,Plus,Minus,credit_card,Tax

12-9321320,Pickle,Réjane,,$41564.80,$46160.22,3557725016917448,$16408.20

14-4213773,Hartford,Léa,,$46310.29,$13116.21,3544948067779642,$13397.19

92-0579703,Caldecott,Adélaïde,lcaldecott2@com.com,$44330.64,$9725.39,4508863396652307,$35768.27

43-3066209,Dohrmann,Cloé,odohrmann3@gov.uk,$32271.10,$33467.96,6398709109212517,$3876.95

39-1753452,Pes,Zhì,dpes4@trellian.com,$24467.01,$5220.55,3562917312487449,$5559.11

98-1640163,Satyford,Pénélope,asatyford5@friendfeed.com,$31883.05,$11529.92,3564171803659243,$5939.26

58-9143184,Corballis,Clémentine,scorballis6@mapquest.com,$25340.90,$3241.86,341075109903237,$45625.23

87-3594410,Hiseman,Bénédicte,chiseman7@sciencedirect.com,$46081.76,$36954.21,3548420880126084,$19464.18



Строки перебираются, но теперь необходимо разделить единую строку на отдельные элементы данных. Для этого выполним метод `split(sep)`, который разрезает строку, используя символ `sep` в качестве разделителя. Также обратим внимание на тот факт, что в первой строке содержатся не данные, а заголовки столбцов.

In [7]:
# доработка предыдущего кода
F=open('data_v3.csv')
header = F.readline().strip().split(',')
print(header)
print('-'*20)
n_to_print=10
for line in F:
  # производим обработку в любом случае ...
  data = line.strip().split(',')
  # ... аа печатаем только первые n_to_print
  n_to_print-=1
  if n_to_print>0:
    print(data)

['ID', 'LastName', 'FirstName', 'email', 'Plus', 'Minus', 'credit_card', 'Tax']
--------------------
['12-9321320', 'Pickle', 'Réjane', '', '$41564.80', '$46160.22', '3557725016917448', '$16408.20']
['14-4213773', 'Hartford', 'Léa', '', '$46310.29', '$13116.21', '3544948067779642', '$13397.19']
['92-0579703', 'Caldecott', 'Adélaïde', 'lcaldecott2@com.com', '$44330.64', '$9725.39', '4508863396652307', '$35768.27']
['43-3066209', 'Dohrmann', 'Cloé', 'odohrmann3@gov.uk', '$32271.10', '$33467.96', '6398709109212517', '$3876.95']
['39-1753452', 'Pes', 'Zhì', 'dpes4@trellian.com', '$24467.01', '$5220.55', '3562917312487449', '$5559.11']
['98-1640163', 'Satyford', 'Pénélope', 'asatyford5@friendfeed.com', '$31883.05', '$11529.92', '3564171803659243', '$5939.26']
['58-9143184', 'Corballis', 'Clémentine', 'scorballis6@mapquest.com', '$25340.90', '$3241.86', '341075109903237', '$45625.23']
['87-3594410', 'Hiseman', 'Bénédicte', 'chiseman7@sciencedirect.com', '$46081.76', '$36954.21', '35484208801

Метод `split()` вернул список отдельных строк - полстрок исходной строки. Теперь необходимо сформировать структуру данных максимально полно описывающую предметную область.
Для этого сформируем список словарей, то есть все данные в процессе чтения будем построчно перемещать в список `all_clients`, при этом каждый клиент будет представлен в виде словаря. Например первая строка в файле будет выглядеть следующим образом:

`example = {'ID' : '12-9321320', 'LastName':'Pickle', 'FirstName':'Réjane', 'email':None, 'Plus':41564.80, 'Minus':46160.22, 'credit_card':'3557725016917448', 'Tax':16408.20}`

то есть будет сформирована структура данных в которой все элементыы представляют собой пары вида `ключ` : `значение`, при этом как и в списке возможно обращение к элементамм данной структуры по индексу, только в словарях вместо целочисленного индекса используется ключ.
Например, применительно к первой строке, обратиться к сумме поступления для ее печати можно следующим образом:

`print( example['Plus'] )`

In [4]:
# Доработаем предыдущий скрипт
F=open('data_v3.csv')
header = F.readline().strip().split(',')

# список всех клиентов
all_clients=[]

for line in F:
  data = line.strip().split(',')
  client = {} # создаем пустой словарь
  for i, key in enumerate(header): # обходим все заголовки
    client[key] = data[i] # добавляем в словарь клюс:значение

  # некоторые значения придется конвертировать к подходящим типам
  client['Minus'] = float(client['Minus'][1:])
  client['Plus'] = float(client['Plus'][1:])
  client['Tax'] = float(client['Tax'][1:])

  all_clients.append(client) # добавляем клиента в скписок

print(all_clients[:10]) # напечатаем первых 10 клиентов

F.close() # файл больше не нужен

[{'ID': '12-9321320', 'LastName': 'Pickle', 'FirstName': 'Réjane', 'email': '', 'Plus': 41564.8, 'Minus': 46160.22, 'credit_card': '3557725016917448', 'Tax': 16408.2}, {'ID': '14-4213773', 'LastName': 'Hartford', 'FirstName': 'Léa', 'email': '', 'Plus': 46310.29, 'Minus': 13116.21, 'credit_card': '3544948067779642', 'Tax': 13397.19}, {'ID': '92-0579703', 'LastName': 'Caldecott', 'FirstName': 'Adélaïde', 'email': 'lcaldecott2@com.com', 'Plus': 44330.64, 'Minus': 9725.39, 'credit_card': '4508863396652307', 'Tax': 35768.27}, {'ID': '43-3066209', 'LastName': 'Dohrmann', 'FirstName': 'Cloé', 'email': 'odohrmann3@gov.uk', 'Plus': 32271.1, 'Minus': 33467.96, 'credit_card': '6398709109212517', 'Tax': 3876.95}, {'ID': '39-1753452', 'LastName': 'Pes', 'FirstName': 'Zhì', 'email': 'dpes4@trellian.com', 'Plus': 24467.01, 'Minus': 5220.55, 'credit_card': '3562917312487449', 'Tax': 5559.11}, {'ID': '98-1640163', 'LastName': 'Satyford', 'FirstName': 'Pénélope', 'email': 'asatyford5@friendfeed.com', '

## Решение задачи 1

Первая задача:
Найдите держателя карты с максимальной прибылью. В ответе необходимо указать ФИО держателя и максимальную прибыль (прибыль = поступления – списания).

Переберем все элементы в списке `all_clients` и выберем клиента с максимальной разницей `client['Plus'] - client['Minus']`



In [5]:
# переменные для хранения целевого клиента и максимальной прибыли
target_client=[]
max_profit=0

for client in all_clients:
  if client['Plus'] - client['Minus'] > max_profit:
    max_profit = client['Plus'] - client['Minus']
    target_client = client

print('Максимальная прибыль: ', max_profit)
print('Клиент с максимальной прибылью', target_client)

Максимальная прибыль:  45913.590000000004
Клиент с максимальной прибылью {'ID': '98-9398689', 'LastName': 'Lingwood', 'FirstName': 'Maïté', 'email': 'mlingwood1y@wordpress.com', 'Plus': 47412.86, 'Minus': 1499.27, 'credit_card': '201561279094179', 'Tax': 23816.11}


## Решение задачи 2

Вторая задача: Найдите количество держателей карт, у которых сумма уплаченного налога больше среднего арифметического значения налога по всей выборке.

Задачу можно решить также перебором всей выборки, но предварительно необходимо определить среднее значение уплаченных налогов

In [6]:
# Определим среднее значение нолога
total_tax = 0
count_tax = len(all_clients)

for client in all_clients:
  total_tax += client['Tax']

avg_tax = total_tax / count_tax

print('Среднее значение налога = %.2f' % avg_tax)

counter=0
for client in all_clients:
  if client['Tax'] > avg_tax:
    counter+=1

print('Количество клиентов с налогом выше среднего: ', counter)

Среднее значение налога = 24939.93
Количество клиентов с налогом выше среднего:  489


В данной работе были использованы операторы `if` (условный оператор), `for ... in range(...)` (оператор цикла с обходом диапазона). Применялись функции `print()`, `len()`, `max()`. В работе использовался тип данных `dict()`

# Контрольные вопросы

Ответы на контрольные вопросы с кодом ....